File: calculadora_cientifica_gui.py
#!/usr/bin/env python3
"""
Calculadora Científica Python
Suporte a cálculos trigonométricos, logaritmos, potenciação, raiz e constantes.
Execução: python calculadora_cientifica.py
"""

import tkinter as tk
from tkinter import ttk
import math


class CalculadoraCientifica:
    def __init__(self, root):
        self.root = root
        self.root.title("Calculadora Científica Python")
        self.root.geometry("460x580")
        self.root.resizable(False, False)
        self.root.configure(bg="#0f172a")

        self.graus = True  # Modo Graus ou Radianos
        self.memoria = 0.0
        self.historico = []

        self._construir_interface()

    def _construir_interface(self):
        # Top bar: Modo Ângulo & Memória
        topo = tk.Frame(self.root, bg="#0f172a")
        topo.pack(fill=tk.X, padx=16, pady=(12, 4))

        self.btn_modo_ang = tk.Button(
            topo,
            text="DEG",
            font=("Helvetica", 10, "bold"),
            bg="#1e293b",
            fg="#38bdf8",
            bd=0,
            padx=10,
            pady=2,
            cursor="hand2",
            command=self.alternar_angulo
        )
        self.btn_modo_ang.pack(side=tk.LEFT)

        self.lbl_mem = tk.Label(
            topo,
            text="M: 0",
            font=("Helvetica", 9),
            bg="#0f172a",
            fg="#94a3b8"
        )
        self.lbl_mem.pack(side=tk.RIGHT)

        # Display de Fórmula e Resultado
        display_box = tk.Frame(self.root, bg="#020617", bd=1, relief=tk.SOLID)
        display_box.pack(fill=tk.X, padx=16, pady=6)

        self.lbl_expressao = tk.Label(
            display_box,
            text="",
            font=("Helvetica", 11),
            fg="#64748b",
            bg="#020617",
            anchor="e",
            padx=12,
            pady=4
        )
        self.lbl_expressao.pack(fill=tk.X)

        self.display = tk.Entry(
            display_box,
            font=("Helvetica", 24, "bold"),
            justify="right",
            bd=0,
            bg="#020617",
            fg="#f8fafc",
            insertbackground="#f8fafc"
        )
        self.display.pack(fill=tk.X, padx=8, pady=(0, 10))
        self.display.insert(0, "0")

        # Grade de Botões Científicos
        grid_frame = tk.Frame(self.root, bg="#0f172a")
        grid_frame.pack(fill=tk.BOTH, expand=True, padx=16, pady=(6, 16))

        for r in range(7):
            grid_frame.rowconfigure(r, weight=1)
        for c in range(5):
            grid_frame.columnconfigure(c, weight=1)

        botoes = [
            # Linha 0: Memória & Constantes
            ("MC", 0, 0, "#1e293b", "#94a3b8", self.mem_clear),
            ("MR", 0, 1, "#1e293b", "#94a3b8", self.mem_recall),
            ("M+", 0, 2, "#1e293b", "#94a3b8", self.mem_add),
            ("π", 0, 3, "#1e293b", "#38bdf8", lambda: self.inserir_constante(math.pi)),
            ("e", 0, 4, "#1e293b", "#38bdf8", lambda: self.inserir_constante(math.e)),

            # Linha 1: Trigonometria & Raiz
            ("sin", 1, 0, "#1e293b", "#38bdf8", lambda: self.funcao_unaria("sin")),
            ("cos", 1, 1, "#1e293b", "#38bdf8", lambda: self.funcao_unaria("cos")),
            ("tan", 1, 2, "#1e293b", "#38bdf8", lambda: self.funcao_unaria("tan")),
            ("√", 1, 3, "#1e293b", "#38bdf8", lambda: self.funcao_unaria("sqrt")),
            ("x²", 1, 4, "#1e293b", "#38bdf8", lambda: self.funcao_unaria("sqr")),

            # Linha 2: Logaritmos & Potência
            ("ln", 2, 0, "#1e293b", "#38bdf8", lambda: self.funcao_unaria("ln")),
            ("log", 2, 1, "#1e293b", "#38bdf8", lambda: self.funcao_unaria("log10")),
            ("xʸ", 2, 2, "#1e293b", "#38bdf8", lambda: self.inserir_operador("**")),
            ("n!", 2, 3, "#1e293b", "#38bdf8", lambda: self.funcao_unaria("fact")),
            ("1/x", 2, 4, "#1e293b", "#38bdf8", lambda: self.funcao_unaria("inv")),

            # Linha 3: Controles & Divisão
            ("C", 3, 0, "#334155", "#f87171", self.limpar_tudo),
            ("CE", 3, 1, "#334155", "#cbd5e1", self.limpar_entrada),
            ("(", 3, 2, "#334155", "#cbd5e1", lambda: self.inserir("(")),
            (")", 3, 3, "#334155", "#cbd5e1", lambda: self.inserir(")")),
            ("÷", 3, 4, "#0284c7", "#ffffff", lambda: self.inserir_operador("/")),

            # Linha 4: 7, 8, 9 & Multiplicação
            ("7", 4, 0, "#1e293b", "#f8fafc", lambda: self.inserir("7")),
            ("8", 4, 1, "#1e293b", "#f8fafc", lambda: self.inserir("8")),
            ("9", 4, 2, "#1e293b", "#f8fafc", lambda: self.inserir("9")),
            ("%", 4, 3, "#334155", "#cbd5e1", lambda: self.funcao_unaria("pct")),
            ("×", 4, 4, "#0284c7", "#ffffff", lambda: self.inserir_operador("*")),

            # Linha 5: 4, 5, 6 & Subtração
            ("4", 5, 0, "#1e293b", "#f8fafc", lambda: self.inserir("4")),
            ("5", 5, 1, "#1e293b", "#f8fafc", lambda: self.inserir("5")),
            ("6", 5, 2, "#1e293b", "#f8fafc", lambda: self.inserir("6")),
            ("+/-", 5, 3, "#334155", "#cbd5e1", self.inverter_sinal),
            ("-", 5, 4, "#0284c7", "#ffffff", lambda: self.inserir_operador("-")),

            # Linha 6: 1, 2, 3, 0, ., = & Adição
            ("1", 6, 0, "#1e293b", "#f8fafc", lambda: self.inserir("1")),
            ("2", 6, 1, "#1e293b", "#f8fafc", lambda: self.inserir("2")),
            ("3", 6, 2, "#1e293b", "#f8fafc", lambda: self.inserir("3")),
            ("0", 7, 0, "#1e293b", "#f8fafc", lambda: self.inserir("0")),
            (".", 7, 1, "#1e293b", "#f8fafc", lambda: self.inserir(".")),
            ("=", 7, 2, "#059669", "#ffffff", self.calcular, 2),
            ("+", 6, 4, "#0284c7", "#ffffff", lambda: self.inserir_operador("+")),
        ]

        for item in botoes:
            txt, r, c, bg_c, fg_c, cmd = item[:6]
            cspan = item[6] if len(item) > 6 else 1
            btn = tk.Button(
                grid_frame,
                text=txt,
                font=("Helvetica", 11, "bold"),
                bg=bg_c,
                fg=fg_c,
                activebackground=bg_c,
                activeforeground=fg_c,
                bd=0,
                cursor="hand2",
                command=cmd
            )
            btn.grid(row=r, column=c, columnspan=cspan, sticky="nsew", padx=2, pady=2)

    def alternar_angulo(self):
        self.graus = not self.graus
        self.btn_modo_ang.config(text="DEG" if self.graus else "RAD")

    def inserir(self, char):
        cur = self.display.get()
        if cur == "0" and char not in ".":
            self.display.delete(0, tk.END)
            self.display.insert(tk.END, char)
        elif cur == "Erro":
            self.display.delete(0, tk.END)
            self.display.insert(tk.END, char)
        else:
            self.display.insert(tk.END, char)

    def inserir_constante(self, val):
        self.display.delete(0, tk.END)
        self.display.insert(0, f"{val:.8g}")

    def limpar_tudo(self):
        self.display.delete(0, tk.END)
        self.display.insert(0, "0")
        self.lbl_expressao.config(text="")

    def limpar_entrada(self):
        self.display.delete(0, tk.END)
        self.display.insert(0, "0")

    def inverter_sinal(self):
        try:
            v = float(self.display.get())
            res = -v
            self.display.delete(0, tk.END)
            self.display.insert(0, f"{int(res) if res.is_integer() else res}")
        except ValueError:
            pass

    def inserir_operador(self, op):
        cur = self.display.get()
        self.lbl_expressao.config(text=f"{self.lbl_expressao['text']} {cur} {op}")
        self.display.delete(0, tk.END)
        self.display.insert(0, "0")

    def funcao_unaria(self, nome):
        try:
            x = float(self.display.get())
            if nome == "sin":
                rad = math.radians(x) if self.graus else x
                res = math.sin(rad)
            elif nome == "cos":
                rad = math.radians(x) if self.graus else x
                res = math.cos(rad)
            elif nome == "tan":
                rad = math.radians(x) if self.graus else x
                res = math.tan(rad)
            elif nome == "sqrt":
                if x < 0:
                    raise ValueError("Raiz de número negativo")
                res = math.sqrt(x)
            elif nome == "sqr":
                res = x ** 2
            elif nome == "ln":
                res = math.log(x)
            elif nome == "log10":
                res = math.log10(x)
            elif nome == "fact":
                if x < 0 or not x.is_integer():
                    raise ValueError("Fatorial requer inteiro >= 0")
                res = math.factorial(int(x))
            elif nome == "inv":
                res = 1 / x
            elif nome == "pct":
                res = x / 100.0
            else:
                return

            if isinstance(res, float):
                res = round(res, 8)
                if res.is_integer():
                    res = int(res)

            self.display.delete(0, tk.END)
            self.display.insert(0, str(res))
        except Exception as err:
            self.display.delete(0, tk.END)
            self.display.insert(0, "Erro")

    def mem_clear(self):
        self.memoria = 0.0
        self.lbl_mem.config(text="M: 0")

    def mem_recall(self):
        self.display.delete(0, tk.END)
        v = self.memoria
        self.display.insert(0, f"{int(v) if v.is_integer() else v}")

    def mem_add(self):
        try:
            self.memoria += float(self.display.get())
            self.lbl_mem.config(text=f"M: {self.memoria:.3g}")
        except ValueError:
            pass

    def calcular(self):
        formula = f"{self.lbl_expressao['text']} {self.display.get()}".strip()
        if not formula:
            return
        try:
            # Avaliação com escopo matemático seguro
            escopo = {
                "math": math,
                "pi": math.pi,
                "e": math.e,
                "sin": math.sin,
                "cos": math.cos,
                "tan": math.tan,
                "sqrt": math.sqrt
            }
            res = eval(formula, {"__builtins__": {}}, escopo)
            if isinstance(res, float):
                res = round(res, 8)
                if res.is_integer():
                    res = int(res)
            self.lbl_expressao.config(text=f"{formula} =")
            self.display.delete(0, tk.END)
            self.display.insert(0, str(res))
        except ZeroDivisionError:
            self.display.delete(0, tk.END)
            self.display.insert(0, "Divisão por Zero")
        except Exception:
            self.display.delete(0, tk.END)
            self.display.insert(0, "Erro de Sintaxe")


if __name__ == "__main__":
    janela = tk.Tk()
    app = CalculadoraCientifica(janela)
    janela.mainloop()

File: CalculadoraCientificaGUI.java
import javax.swing.*;
import java.awt.*;
import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;
import java.util.Locale;

/**
 * CalculadoraCientificaGUI.java
 * Interface Gráfica Científica em Java Swing com Funções Trigonométricas, Logaritmos e Memória.
 * Execução: java CalculadoraCientificaGUI.java
 */
public class CalculadoraCientificaGUI extends JFrame implements ActionListener {

    private final JTextField visor;
    private final JLabel labelStatus;
    private double primeiroNumero = 0;
    private String operador = "";
    private boolean inicioNovo = true;
    private boolean modoRadianos = false;
    private double memoria = 0;
    private final int precisao = 6;

    public CalculadoraCientificaGUI() {
        super("Calculadora Java - Científica");
        setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);
        setSize(480, 560);
        setLocationRelativeTo(null);
        setResizable(false);

        Color corFundo = new Color(24, 24, 27);
        getContentPane().setBackground(corFundo);
        setLayout(new BorderLayout(8, 8));

        // Painel Superior
        JPanel topo = new JPanel(new BorderLayout(4, 4));
        topo.setBackground(corFundo);
        topo.setBorder(BorderFactory.createEmptyBorder(12, 14, 4, 14));

        labelStatus = new JLabel("MODO: GRAUS (DEG) | M = 0", SwingConstants.RIGHT);
        labelStatus.setFont(new Font("SansSerif", Font.PLAIN, 12));
        labelStatus.setForeground(new Color(161, 161, 170));
        topo.add(labelStatus, BorderLayout.NORTH);

        visor = new JTextField("0");
        visor.setEditable(false);
        visor.setHorizontalAlignment(JTextField.RIGHT);
        visor.setFont(new Font("SansSerif", Font.BOLD, 28));
        visor.setBackground(new Color(9, 9, 11));
        visor.setForeground(Color.WHITE);
        visor.setBorder(BorderFactory.createCompoundBorder(
            BorderFactory.createLineBorder(new Color(63, 63, 70)),
            BorderFactory.createEmptyBorder(8, 10, 8, 10)
        ));
        topo.add(visor, BorderLayout.CENTER);
        add(topo, BorderLayout.NORTH);

        // Grade de Botões
        JPanel grid = new JPanel(new GridLayout(7, 5, 6, 6));
        grid.setBackground(corFundo);
        grid.setBorder(BorderFactory.createEmptyBorder(4, 14, 14, 14));

        String[] botoes = {
            "DEG/RAD", "MC", "MR", "M+", "M-",
            "sin", "cos", "tan", "π", "e",
            "ln", "log", "x²", "xʸ", "√",
            "C", "⌫", "+/-", "%", "÷",
            "7", "8", "9", "n!", "×",
            "4", "5", "6", "1/x", "-",
            "1", "2", "3", "0", "+",
            ".", "EXP", "(", ")", "="
        };

        for (String b : botoes) {
            JButton btn = new JButton(b);
            btn.setFont(new Font("SansSerif", Font.BOLD, 13));
            btn.setFocusPainted(false);
            btn.setCursor(new Cursor(Cursor.HAND_CURSOR));

            if ("=÷×-+".contains(b)) {
                btn.setBackground(new Color(249, 115, 22));
                btn.setForeground(Color.WHITE);
            } else if ("sin cos tan ln log x² xʸ √ n! 1/x π e DEG/RAD".contains(b)) {
                btn.setBackground(new Color(30, 41, 59));
                btn.setForeground(new Color(56, 189, 248));
            } else if ("MC MR M+ M- C ⌫ +/- %".contains(b)) {
                btn.setBackground(new Color(63, 63, 70));
                btn.setForeground(Color.WHITE);
            } else {
                btn.setBackground(new Color(39, 39, 42));
                btn.setForeground(Color.WHITE);
            }

            btn.addActionListener(this);
            grid.add(btn);
        }

        add(grid, BorderLayout.CENTER);
    }

    @Override
    public void actionPerformed(ActionEvent e) {
        String cmd = e.getActionCommand();

        try {
            if (cmd.matches("[0-9]")) {
                if (inicioNovo || visor.getText().equals("0")) {
                    visor.setText(cmd);
                    inicioNovo = false;
                } else {
                    visor.setText(visor.getText() + cmd);
                }
            } else if (cmd.equals(".")) {
                if (inicioNovo) { visor.setText("0."); inicioNovo = false; }
                else if (!visor.getText().contains(".")) visor.setText(visor.getText() + ".");
            } else if (cmd.equals("C")) {
                visor.setText("0");
                primeiroNumero = 0;
                operador = "";
                inicioNovo = true;
            } else if (cmd.equals("⌫")) {
                String t = visor.getText();
                visor.setText(t.length() > 1 && !inicioNovo ? t.substring(0, t.length() - 1) : "0");
            } else if (cmd.equals("DEG/RAD")) {
                modoRadianos = !modoRadianos;
                atualizarStatus();
            } else if (cmd.equals("π")) {
                visor.setText(formatar(Math.PI));
                inicioNovo = true;
            } else if (cmd.equals("e")) {
                visor.setText(formatar(Math.E));
                inicioNovo = true;
            } else if (cmd.equals("M+")) {
                memoria += Double.parseDouble(visor.getText());
                atualizarStatus();
                inicioNovo = true;
            } else if (cmd.equals("M-")) {
                memoria -= Double.parseDouble(visor.getText());
                atualizarStatus();
                inicioNovo = true;
            } else if (cmd.equals("MR")) {
                visor.setText(formatar(memoria));
                inicioNovo = true;
            } else if (cmd.equals("MC")) {
                memoria = 0;
                atualizarStatus();
            } else if (cmd.equals("sin") || cmd.equals("cos") || cmd.equals("tan")) {
                double val = Double.parseDouble(visor.getText());
                double angulo = modoRadianos ? val : Math.toRadians(val);
                double res = cmd.equals("sin") ? Math.sin(angulo) : cmd.equals("cos") ? Math.cos(angulo) : Math.tan(angulo);
                visor.setText(formatar(res));
                inicioNovo = true;
            } else if (cmd.equals("√")) {
                double v = Double.parseDouble(visor.getText());
                if (v < 0) visor.setText("Erro: Raiz < 0");
                else visor.setText(formatar(Math.sqrt(v)));
                inicioNovo = true;
            } else if (cmd.equals("x²")) {
                double v = Double.parseDouble(visor.getText());
                visor.setText(formatar(v * v));
                inicioNovo = true;
            } else if (cmd.equals("ln")) {
                double v = Double.parseDouble(visor.getText());
                if (v <= 0) visor.setText("Erro: ln <= 0");
                else visor.setText(formatar(Math.log(v)));
                inicioNovo = true;
            } else if (cmd.equals("log")) {
                double v = Double.parseDouble(visor.getText());
                if (v <= 0) visor.setText("Erro: log <= 0");
                else visor.setText(formatar(Math.log10(v)));
                inicioNovo = true;
            } else if (cmd.equals("1/x")) {
                double v = Double.parseDouble(visor.getText());
                if (v == 0) visor.setText("Erro: Div/0");
                else visor.setText(formatar(1.0 / v));
                inicioNovo = true;
            } else if (cmd.equals("n!")) {
                double v = Double.parseDouble(visor.getText());
                if (v < 0 || v != (int) v) visor.setText("Erro: Inteiro >= 0");
                else {
                    long f = 1;
                    for (int i = 2; i <= (int) v; i++) f *= i;
                    visor.setText(String.valueOf(f));
                }
                inicioNovo = true;
            } else if ("+-×÷xʸ".contains(cmd)) {
                primeiroNumero = Double.parseDouble(visor.getText());
                operador = cmd;
                inicioNovo = true;
            } else if (cmd.equals("=")) {
                if (!operador.isEmpty()) {
                    double n2 = Double.parseDouble(visor.getText());
                    double r = 0;
                    switch (operador) {
                        case "+": r = primeiroNumero + n2; break;
                        case "-": r = primeiroNumero - n2; break;
                        case "×": r = primeiroNumero * n2; break;
                        case "÷":
                            if (n2 == 0) { visor.setText("Erro: Div/0"); return; }
                            r = primeiroNumero / n2; break;
                        case "xʸ": r = Math.pow(primeiroNumero, n2); break;
                    }
                    visor.setText(formatar(r));
                    operador = "";
                    inicioNovo = true;
                }
            }
        } catch (Exception ex) {
            visor.setText("Erro");
            inicioNovo = true;
        }
    }

    private void atualizarStatus() {
        labelStatus.setText(String.format("MODO: %s | M = %s",
            modoRadianos ? "RADIANOS (RAD)" : "GRAUS (DEG)", formatar(memoria)));
    }

    private String formatar(double val) {
        if (Double.isNaN(val) || Double.isInfinite(val)) return "Erro";
        if (val == (long) val) return String.valueOf((long) val);
        return String.format(Locale.US, "%." + precisao + "f", val).replaceAll("0+$", "").replaceAll("\\.$", "");
    }

    public static void main(String[] args) {
        SwingUtilities.invokeLater(() -> new CalculadoraCientificaGUI().setVisible(true));
    }
}
